### PR TITLE
Add support for flashing AMD dGPUs

### PR DESCRIPTION
**Pull request by superm1**
Type of pull request:

- [ ] New plugin (Please include [new plugin checklist](https://github.com/fwupd/fwupd/wiki/New-plugin-checklist))
- [ ] Code fix
- [x] Feature
- [ ] Documentation

### CODE REVIEW

File: plugins/amd-gpu/fu-amd-gpu-atom-firmware.c [hunk 1]
```diff
@@ -0,0 +1,424 @@
...
+	val = fu_struct_vbios_date_get_year(st);
+	if (!fu_memcpy_safe((guint8 *)vbios_date,
+			    sizeof(vbios_date),
+			    2,
+			    (guint8 *)&val,
+			    sizeof(val),
+			    0,
+			    sizeof(val),
+			    error))
+		return FALSE;
```
Comment (hughsie) [inline plugins/amd-gpu/fu-amd-gpu-atom-firmware.c:158]: wouldn't all this be much simpler and safer if it was just `g_strdup_printf("20%02X/%02X/%02X %02X:%02X", fu_struct_vbios_date_get_year(st), ....)`

File: plugins/amd-gpu/fu-amd-gpu-atom-firmware.c [hunk 2]
```diff
@@ -0,0 +1,424 @@
...
+				    base + idx,
+				    len,
+				    error)) {
+			g_prefix_error(error, "failed to copy string %d: ", i);
+			return FALSE;
+		}
+
+		switch (i) {
+		case FU_ATOM_STRING_INDEX_PART_NUMBER:
+			self->part_number = g_strndup((gchar *)tmp, len);
```
Comment (hughsie) [inline plugins/amd-gpu/fu-amd-gpu-atom-firmware.c:269]: could we use g_strsafe here instead to avoid non-ascii chars?

Comment (hughsie) [reply plugins/amd-gpu/fu-amd-gpu-atom-firmware.c:269]: or fu_memstrsafe perhaps?

File: plugins/amd-gpu/fu-amd-gpu-atom-firmware.c [hunk 3]
```diff
@@ -0,0 +1,424 @@
...
+	/* skip the following 2 chars: 0x0D 0x0A */
+	idx += 2;
+
+	/* get the model number */
+	len = strnlen((gchar *)buf + base + idx, STRLEN_LONG - 1);
+	g_debug("base 0x%lx, idx 0x%lx, len 0x%lx", base, idx, len);
+	if (!fu_memcpy_safe((guint8 *)tmp, sizeof(tmp), 0x0, buf, bufsz, base + idx, len, error)) {
+		g_prefix_error(error, "failed to copy model string: ");
+		return FALSE;
+	}
```
Comment (hughsie) [inline plugins/amd-gpu/fu-amd-gpu-atom-firmware.c:309]: fu_memstrsafe perhaps?

File: plugins/amd-gpu/fu-amd-gpu-atom-firmware.c [hunk 4]
```diff
@@ -0,0 +1,424 @@
...
+	guint16 idx;
+	gsize len;
+
+	base = fu_firmware_get_addr(FU_FIRMWARE(self));
+	idx = fu_struct_atom_rom21_header_get_config_filename_offset(atom_header);
+	len = strnlen((gchar *)(buf + base + idx), STRLEN_NORMAL - 1);
+	if (!fu_memcpy_safe((guint8 *)tmp, sizeof(tmp), 0x0, buf, bufsz, base + idx, len, error)) {
+		g_prefix_error(error, "failed to copy config string: ");
+		return FALSE;
+	}
```
Comment (hughsie) [inline plugins/amd-gpu/fu-amd-gpu-atom-firmware.c:333]: fu_memstrsafe, but maybe you want to wrap it in a static function that includes the strstrip

File: plugins/amd-gpu/fu-amd-gpu-device.c [hunk 5]
```diff
@@ -1,45 +1,275 @@
...
+
+	ddir = g_build_filename(base, "drm", NULL);
+	dir = g_dir_open(ddir, 0, error);
+	if (dir == NULL)
+		return FALSE;
+	while ((f = g_dir_read_name(dir))) {
+		if (g_str_has_prefix(f, "card")) {
+			g_autofree gchar *device_file = NULL;
+
+			device_file = g_build_filename("/", "dev", "dri", f, NULL);
```
Comment (hughsie) [inline plugins/amd-gpu/fu-amd-gpu-device.c:53]: does this need a FU_PATH getter?

File: plugins/amd-gpu/fu-amd-gpu-device.c [hunk 6]
```diff
@@ -1,45 +1,275 @@
...
+	g_autofree gchar *part = NULL;
+
+	if (!fu_udev_device_ioctl(FU_UDEV_DEVICE(device),
+				  DRM_IOCTL_AMDGPU_INFO,
+				  (void *)&request,
+				  NULL,
+				  1000,
+				  error))
+		return FALSE;
+	self->vbios_pn = g_strdup((const gchar *)vbios_info.vbios_pn);
```
Comment (hughsie) [inline plugins/amd-gpu/fu-amd-gpu-device.c:129]: definitely fu_strsafe here.

File: plugins/amd-gpu/fu-amd-gpu-device.c [hunk 7]
```diff
@@ -1,45 +1,275 @@
...
+	part = g_strdup_printf("AMD\\%s", self->vbios_pn);
+	fu_device_add_instance_id(device, part);
+	fu_device_set_version_raw(device, vbios_info.version);
+	fu_device_set_version(device, (const char *)vbios_info.vbios_ver_str);
```
Comment (hughsie) [inline plugins/amd-gpu/fu-amd-gpu-device.c:133]: and here

File: plugins/amd-gpu/fu-amd-gpu-device.c [hunk 8]
```diff
@@ -1,45 +1,275 @@
...
+
+	if (!fu_device_has_flag(device, FWUPD_DEVICE_FLAG_INTERNAL))
+		fu_device_set_name(device, (const char *)vbios_info.name);
```
Comment (hughsie) [inline plugins/amd-gpu/fu-amd-gpu-device.c:136]: and here

File: plugins/amd-gpu/fu-amd-gpu-device.c [hunk 9]
```diff
@@ -1,45 +1,275 @@
...
+			break;
+		fu_device_sleep(device, 100);
+	}
+	if (retries == PSPVBFLASH_MAX_POLL) {
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_WRITE,
+			    "timeout waiting for pspvbflash to complete");
 		return FALSE;
 	}
```
Comment (hughsie) [inline plugins/amd-gpu/fu-amd-gpu-device.c:237]: fu_device_retry_full might be more elegant here.

File: plugins/amd-gpu/fu-amd-gpu-psp-firmware.c [hunk 10]
```diff
@@ -0,0 +1,255 @@
...
+
+#include "fu-amd-gpu-atom-firmware.h"
+#include "fu-amd-gpu-psp-firmware.h"
+#include "fu-amd-gpu-psp-struct.h"
+
+#define PARTITION_A_L2_FWID 0x014D
+#define PARTITION_B_L2_FWID 0x014E
+#define ISH_A_FWID	    0x013C
+#define ISH_B_FWID	    0x013D
+#define ATOM_CSM_FWID	    0x1
```
Comment (hughsie) [inline plugins/amd-gpu/fu-amd-gpu-psp-firmware.c:26]: maybe an enum in the .rs file?

**Comment by hughsie** [comment]
@superm1 I've pushed some fixups, I hope that's okay.

**Comment by superm1** [comment]
I've squashed your fixups (thanks!) and also added fixes for everything except the `fu_device_retry`.  Can you push what you have in mind for that?

**Comment by superm1** [comment]
@valentindavid can you double check if snapd's interface needs anything new for the IOCTLs to `/dev/dri/card*`?

**Comment by superm1** [comment]
> @valentindavid can you double check if snapd's interface needs anything new for the IOCTLs to `/dev/dri/card*`?

Actually I think I found what was needed - the `opengl` plug.  So I added it above.  If that's wrong, let me know.

File: plugins/amd-gpu/fu-amd-gpu-device.c (hunk 9 above)
Comment (hughsie) [reply plugins/amd-gpu/fu-amd-gpu-device.c]: @superm1 i've pushed a fixup, can you check that makes sense pls. then LGTM.

**Comment by valentindavid** [comment]
> > @valentindavid can you double check if snapd's interface needs anything new for the IOCTLs to `/dev/dri/card*`?
> 
> Actually I think I found what was needed - the `opengl` plug. So I added it above. If that's wrong, let me know.

I think it is fine.

**Comment by superm1** [comment]
> > > @valentindavid can you double check if snapd's interface needs anything new for the IOCTLs to `/dev/dri/card*`?
> > 
> > Actually I think I found what was needed - the `opengl` plug. So I added it above. If that's wrong, let me know.
> 
> I think it is fine.

Thanks, what about the new sysfs attributes that get written to?

**Comment by valentindavid** [comment]
`/sys/bus/pci/drivers/*/amdgpu`?
I suppose we have to add it in fwupd interface. as:
```
/sys/devices/**/amdgpu rw,
```

**Comment by valentindavid** [comment]
Sorry, it should be `/sys/bus/pci/drivers/amdgpu`. It is a driver not a device.

**Comment by superm1** [comment]
> Sorry, it should be `/sys/bus/pci/drivers/amdgpu`. It is a driver not a device.

It's an attribute that shows up for the device, but will only be present when amdgpu is loaded.

**Comment by valentindavid** [comment]
Also, fwupd is already given `capability sys_admin,`. So the ioctl should be fine.

**Comment by valentindavid** [comment]
Should it be?
```
  /sys/bus/pci/drivers/amdgpu/* w,
```

**Comment by superm1** [comment]
> Should it be?
> 
> ```
>   /sys/bus/pci/drivers/amdgpu/* w,
> ```

I think that's a symlink which you usually can't use.  Maybe just allow those two new sysfs files on all PCI devices?

File: plugins/amd-gpu/fu-amd-gpu-device.c (hunk 9 above)
Comment (superm1) [reply plugins/amd-gpu/fu-amd-gpu-device.c]: thanks!

**Comment by valentindavid** [comment]
> I think that's a symlink which you usually can't use. Maybe just allow those two new sysfs files on all PCI devices?

On my computer:
```
$ realpath -e /sys/bus/pci/drivers/amdgpu
/sys/bus/pci/drivers/amdgpu
```

**Comment by superm1** [comment]
> > I think that's a symlink which you usually can't use. Maybe just allow those two new sysfs files on all PCI devices?
> 
> On my computer:
> 
> ```
> $ realpath -e /sys/bus/pci/drivers/amdgpu
> /sys/bus/pci/drivers/amdgpu
> ```

But the devices within it are symlinks.  Doesn't that cause a problem?

**Comment by superm1** [comment]
> > > I think that's a symlink which you usually can't use. Maybe just allow those two new sysfs files on all PCI devices?
> > 
> > 
> > On my computer:
> > ```
> > $ realpath -e /sys/bus/pci/drivers/amdgpu
> > /sys/bus/pci/drivers/amdgpu
> > ```
> 
> But the devices within it are symlinks. Doesn't that cause a problem?

IE on my local machine:
```
$ realpath /sys/bus/pci/drivers/amdgpu/0000\:61\:00.0/apu_thermal_cap
/sys/devices/pci0000:60/0000:60:03.1/0000:61:00.0/apu_thermal_cap
```

**Comment by valentindavid** [comment]
I misunderstood. Is it any specific attribute we need to match there?

**Comment by superm1** [comment]
Yes
`psp_vbflash` and `psp_vbflash_status`

**Pull request merged**